doc/man1/linxstat.1
.TH LINXSTAT 1 "2006-07-30" 1.0 "LINX"
.SH NAME
linxstat \- display specific information about LINX
.SH SYNOPSIS
.B linxstat [-a] [-b] [-f] [-h] [-H] [-q] [-s] [-S] [-t 
.I type
.B ] [-T] [-l
.I length
.B ] [-n
.I hunt_name
.B ]
.SH DESCRIPTION
.B linxstat
is a utility to display the status of the LINX communication system,
including information about all local and known remote LINX endpoints 
and LINX links to remote nodes. The names, types, status and local 
identifiers are displayed as well as queues, attaches and hunts.
The information is fetched from the LINX kernel module.
The following columns can be output, in addition to a leading 
summary (compare the example):
.br


.I PID
is the id of the process owning the LINX endpoint or LINX link.
As each remote endpoint are represented by a local LINX endpoint
with a local identifier (spid), they are all owned by one 
common local process.
.br

.I spid
is the local LINX binary identifier for the LINX endpoint or LINX link.
.br

.I name
is the huntname of the LINX endpoint or LINX link.
If a name is too long for the column, a "+" sign is displayed
and the tailing part of the name.
.br

.I type
is the type of entry, i.e. 
.IR "local"
for a local LINX endpoint,
.IR "link"
for a LINX link to another node,
.IR "remote"
for a local representation of a remote LINX endpoint,
.IR "zombie"
for a LINX endpoint which has been closed,
.IR "illegal"
if an endpoint structure is illegal.
.IR "unknown"
if linxstat cannot reach the LINX endpoint.
.br

.I state
is the state of the endpoint or link. The main states are 
.IR "running"
not waiting for anything,
.IR "recv"
while waiting to receive signal buffers,
.IR "poll"
while waiting for events with the 
.IR poll "(2)"
or 
.IR select "(2)"
system call,
.IR "unknown"
if linxstat cannot get the state.
.br

.I queue
is the receive queue. This is the number of signal buffers
delivered to the LINX endpoint, but not yet received.
.br

.I attach(from/to)
are the pending attach information. Both the number of attaches to
each LINX endpoint from other endpoints and the number of attaches
from each endpoint to other endpoints are displayed.
.br

.I hunt
are pending hunts owned by each LINX endpoint. 
It also includes pending hunts, sent by others but with this endpoint as sender.
.br

.I tmo
are pending timeouts for each LINX endpoint.

.SH OPTIONS
.IP "-a"
Display more verbose pending attach information.
.IP "-b"
Batch mode, only display the number of open LINX endpoints.
.IP "-f"
Print receive filters.
.IP "-h"
Display a short help message.
.IP "-H"
Display more verbose pending hunt information. 
.IP "-q"
Display more verbose receive queue information.
.IP "-s"
Display less information (simple). Only endpoint-specific information are 
displayed, no summary, no title and no counter columns.
.IP "-S"
If the LINX kernel module has been compiled with per-socket statistics
(SOCK_STAT) then linxstat can display statistics for each socket.
.IP "-t type"
Display only sockets of the specified type.
.IP "-T"
Print pending timeouts.
.IP "-l length"
Specify length of name to be printed, default 18, minimum 5
.IP "-n hunt_name"
Print info only for hunt_name

.SH FILES
None.
.SH DIAGNOSTICS
.I linxstat
writes all information to standard output.
.SH KNOWN BUGS
linxstat queries are not atomic, the state of the system may change
while linxstat is running.

.SH "EXAMPLE"
This is a displayed list from linxstat.

.nf
 local:5 remote:1 link:1
 pending attach:2
 pending hunt:0
 pending timeout:0
 queued signals:0

  PID spid    name               type   state  queue attach(f/t) hunt tmo
24903 0x10003 linx16             link   -      0     1/0         0    0
24903 0x10004 +16/apitest_server remote -      0     0/0         0    0
24909 0x10001 apitest_client     local  run    0     0/1         0    0
24909 0x10024 +est_attach_parent local  poll   0     1/1         0    0
24909 0x10005 not bound          local  run    0     0/0         0    0
24958 0x10023 +test_attach_child local  recv   0     0/0         0    0
24963 0x10025 linxstat           local  run    0     0/0         0    0
.fi

With -a, more information is displayed after each line with 
pending attaches. Extracts from a table, created by linxstat -a: 

.nf
 local:4 remote:1 link:1
 pending attach:1
 pending hunt:0
 pending timeout:0
 queued signals:0

 PID  spid    name               type  state  queue attach(f/t) hunt tmo
25053 0x10002 linx16             link  -      0     1/0         0    0
        Attach info (from):
          [attref:0x00000400   from:0x00010001 signo:0          size: 0]
25053 0x10003 +16/apitest_server remote  -    0     0/0         0    0
25059 0x10001 apitest_client     local run    0     0/1         0    0
        Attach info (to):
          [attref:0x00000400   to:0x00010001 signo:0          size:   0]
25059 0x10004 not bound          local run    0     0/0         0    0
25059 0x10014 +inx_receive_w_tmo local poll   0     0/0         0    0
25083 0x10015 linxstat           local run    0     0/0         0    0
.fi


.SH "SEE ALSO"
.BR linx "(7), "
.BR linx_attach "(3), "
.BR linx_hunt "(3), "
.BR linx_open "(3), "
.BR linxdisc "(8), "
.BR linxdisc.conf "(5), "
.BR linxcfg "(1)"

.SH AUTHOR
Enea LINX team
.SH COPYRIGHT

Copyright (c) 2006-2011, Enea Software AB
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

Redistributions of source code must retain the above copyright notice, this
list of conditions and the following disclaimer.
Redistributions in binary form must reproduce the above copyright notice,
this list of conditions and the following disclaimer in the documentation
and/or other materials provided with the distribution.
Neither the name of Enea Software AB nor the names of its
contributors may be used to endorse or promote products derived from this
software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.


ChangeLog
Version 2.5.1 (2011-07-08)
==========
* The limit of number of sockets per link allowed lowered from 1024 to two.
* Removed a check in LINX kernel module for max signal size.
* Fixes for the multicore header in the ECM and Wireshark patch.
		
Version 2.5.0 (2011-04-26)
==========
* Fixed a bug in linxstat.
* Updated manpage for mkethcon and help info for mkethcon and linxcfg to reflect
  reality.
* libcfg was updated to support cmcl connections.
* Fixed minor bugs in rm*con binaries.
* Fixed build errors for various kernel versions. ipv6 for <2.6.22 and db.c for
  newer ones >2.6.36.
* Fixed some issues in the linx Gateway related to LINXCR-1397
* Added support for ipv6 in the tcp connection manager.
* Improved the performance of intra-process communication with pointer passing.
* Fixed some issues found by coverity.
* Removed some warnings when building linxstat.
* Remove all BUG BUG_ON WARN_ON from linx modules to avoid using __FILE__
  define in the errors. This define causes the linx module to be different
  if compiled in different folder.
* Added fixes in the ECM were a malformed received packet could lead to
  kernel panic.
* Fixed issue with multi-access to LINX sockets from user-space
  applications could lead to kernel panic.
* Added new features to linxstat: print info only for a given huntname
  and display only a specified number of characters from huntname
* Added coreid header
* Added support for 2.5.35 kernels.
* Fixed bug with negative signal filter.
* Fixed bug with missing internal signals in the message trace.
* Added patch file for Ethernet CM to compensate for faulty Ethernet
  drivers that don't allocate enough headroom in skb when using VLAN.
* Added support for 2.6.33 and 2.6.34 kernels.
* Fixed vlan bug reproducing on certain machines
* Fixed bug with receive_w_tmo where tmo=0 resulted in no signal above
  4095 could be received.
* Fixed string handling bug in linxgws.
* Fixed memory allocation issue in linxgws.
* Fixed a bug in linxdisc regarding several interfaces.
* Added ip config in linxnet conf file
* Linxstat now uses a function interface to retrieve info instead of
ioctl.
	
Version 2.4.0 (2010-02-22)
==========	
* Added missing riocm.txt to MANIFEST
* Added RIOCM documentation. net/linx/riocm/riocm.txt
* Updated RELEASE_NOTES for 2.4.0 release.
* Added support for creating links using RapidIO CM connections in linxcfg.

Version 2.4.0-beta (2010-01-15)
==========
* Added a RapidIO connection manager.
* Remove debug printk in device notification handler.

Version 2.3.0 (2009-12-9)
==========
* Fixed build errors for vanilla kernel 2.6.32
* Example conf file for Linxdisc now emphasizes that VLAN interfaces are supported.
* RLNH control data is now always sent from kernel context.
* Altered the keepalive mechanism for the Ethernet CM. The old implementation
  was sensitive to packet loss. Contributed by paull7 @ sourceforge. Note: The
  new implementation is backwards compatible.
* Added support for 2.6.31 kernels.
* Fix LINX_IOCTL_INFO ioctl request for gathering process' receive filters.
* Document LINX_INFO_RECV_QUEUE_2.
* Drop tcp connect packets that doesn't match the current protocol
version.
* Add hardware address size in ethernet connection header.
* Man page for linx_request_new_link and linx_cancel_new_link was missing
* Fixed bug where the request ack bit was not set when the sliding window queue
  is becoming full.
* Fixed memory leaks in the code for create and destroy connections and links.	
* New shared memory connection manager.
* Wireshark plugin now supports LINX TCP Connection manager.
* Wrong RLNH protocol version was listed in documentation.	
	
Version 2.2.0 (2009-03-31)
==========
* New Linxdisc and Linx internal messages patches for Wireshark 1.0.6
* New LINX patches for tcpdump 4.0.0 and libpcap 1.0.0
* Trace of internal LINX messages now use official LINX ethernet type, 0x8911.
* Added new function linx_get_owner(), return the pid of the owner of a LINX endpoint.	
* Added missing files to MANIFEST.
* Example files now have the correct copyright header.
* Fixed oob-issue where the a link in some cases wasn't created.
* Removed a state in the tcp cm state machine.
* Did some major cleanup in tcp cm.
* Removed a kernel function from af_linx.c
* Fixed a bug in tcp cm regarding halt of workqueue when connecting to dead peers
* Linux Gateway Server and Client added to distribution, also gwcmd tool to test Linx Gateway Servers. 
* Removed bug_on's and fixed better error reporting back to user space on failed send.
* Fix bug in linx_info_filter_32
* Fixed alignment issue in linxstat.
* Removed linx_create/destroy_eth/tcp_link functions.
* Avoid calling kmalloc(0) from Linx IPC-layer.

Version 2.1.0 (2008-11-10)
==========	
* LINX now runs on x86 2.6.27 Linux kernel.
* Add support for mips(64).
* Add patches to libpcap and tcpdump configure scripts for Octeon.
* The TCP CM now support OOB messaging.
* Fixes for running w errorchecks enabled.
* Improved garbage collection of sockets in the tcp_cm.
* Minor changes of linxstat output.
* The linxstat program can now tell if a signal in the receive queue is OOB.
* Support for allocating fragmentated skb's is now a LINX kernel module param and no longer
  a comple-time option.
* Fixed bug when building the LINX kernel module with memory fragmentation support on and
  sending large unpaged signals.
* Fixed bug in LINX message trace. If a fragmented message was received, only the first
  fragment was traced.
* Added optimized linx_do_sendmsg call for attach/hunt/tmo/newlink signals.
* Pulled back warning level a bit to avoid complaints from kernel headers.
* Fixed bug that lead to the LINX kernel module couldn't be unloaded.
* linx_skb_copy_bits removed
* linx_ioctl_info_*() functions moved to a seperate file.
* Internal handling of attach/tmo/newlink signals is simplified.
* Updated liblinx to user LINX_IOCTL_SEND and LINX_IOCTL_RECEIVE instead of sendto/recvfrom/sendmsg

linxstat/linxstat.c
/*
 * Copyright (c) 2006-2010, Enea Software AB
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 * Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 * Neither the name of Enea Software AB nor the names of its
 * contributors may be used to endorse or promote products derived from this
 * software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <ctype.h>
#include <errno.h>
#include <getopt.h>

#include <linx.h>
#include <linx_types.h>
#include "linx_info.h"

#define ATTACH_STR_MAX 11

#define DEFAULT_NAME_LENGTH 18
#define MINIMUM_NAME_LENGTH 5
int maxNameLength = DEFAULT_NAME_LENGTH;


char *hunt_name = NULL;


static void usage(void)
{
	printf("Usage: linxstat [-b] [-h] [-aHqSs] [-l <length>] "
		"[-n <hunt_name>] [-t <type>]\n");
	printf("Description: Print LINX information specific socket "
	       "information\n");
	printf("\t-a print verbose pending attach information\n");
	printf("\t-b batch mode, print # of linx end points\n");
	printf("\t-f print receive filters\n");
	printf("\t-H print verbose pending hunt information\n");
	printf("\t-h print this help\n");
	printf("\t-l specify length of name to be printed, default %d, minimum %d\n",
		DEFAULT_NAME_LENGTH, MINIMUM_NAME_LENGTH);
	printf("\t-n <hunt_name> print info only for <hunt_name>\n");
	printf("\t-q print verbose receive queue information\n");
	printf("\t-S print per socket statistics (NOT)\n");
	printf("\t-s print socket individual information only\n");
	printf("\t-t print sockets of specified type only\n");
	printf("\t-T print pending timeouts\n");
	exit(0);
}

static int get_terminal_width(void)
{
	struct winsize win = {0,0,0,0};
	int terminal_width = 0;
	
	ioctl(fileno(stdout), TIOCGWINSZ, &win);
	if (win.ws_col > 0)
		terminal_width = win.ws_col;
	else
		terminal_width = 80;

	return terminal_width;
}

static const char *state_to_str(int state)
{
	switch (state) {
	case LINX_STATE_UNKNOWN:
		return "-";
	case LINX_STATE_RUNNING:
		return "run";
	case LINX_STATE_RECV:
		return "recv";
	case LINX_STATE_POLL:
		return "poll";
	default:
		break;
	}

	return "";
}

static char *strtolower(const char *str)
{
	int i;
	/* 8 characters is enough - we'll only use strtolower to convert
	 * socket types. */
	static char s[8] = { 0, };

	for (i = 0; str[i]; i++)
		s[i] = tolower(str[i]);

	return s;
}

static int get_proc_type(char *str)
{
	char *tmp = strtolower(str);

	if (strcmp("unknown", tmp) == 0) {
		return LINX_TYPE_UNKNOWN;
	} else if (strcmp("local", tmp) == 0) {
		return LINX_TYPE_LOCAL;
	} else if (strcmp("remote", tmp) == 0) {
		return LINX_TYPE_REMOTE;
	} else if (strcmp("link", tmp) == 0) {
		return LINX_TYPE_LINK;
	} else if (strcmp("illegal", tmp) == 0) {
		return LINX_TYPE_ILLEGAL;
	} else if (strcmp("zombie", tmp) == 0) {
		return LINX_TYPE_ZOMBIE;
	} else {
		return LINX_TYPE_UNKNOWN;
	}
}

static const char *type_to_str(int type)
{
	switch (type) {
	case LINX_TYPE_LOCAL:
		return "local";
	case LINX_TYPE_REMOTE:
		return "remote";
	case LINX_TYPE_LINK:
		return "link";
	case LINX_TYPE_ILLEGAL:
		return "illegal";
	case LINX_TYPE_ZOMBIE:
		return "zombie";
	default:
		/* includes LINX_TYPE_UNKNOWN */
		break;
	}

	return "unknown";
}

static void print_attach(int count, struct linx_info_attach *att, int from)
{
	int i;

	printf("\tAttach info (%s):\n",
	       from == LINX_ATTACH_FROM ? "from" : "to");

	for (i = 0; i < count; i++) {
		printf("\t [attref:0x%08x   %s:0x%08x "
		       "signo:%-10d size:%6d]\n",
		       att[i].attref,
		       from == LINX_ATTACH_FROM ? "from" : "to",
		       from == LINX_ATTACH_FROM ?
		       att[i].spid : att[i].attach_signal.from,
		       att[i].attach_signal.signo,
		       att[i].attach_signal.size);
	}
}

static void print_hunt(int count, struct linx_info_hunt *hunt)
{
	int i;

	printf("\tPending hunts (%d, %p):\n", count, (void *)hunt);

	for (i = 0; i < count; i++) {
		printf("\t [signo:%-10d size:%-6d owner:0x%08x"
		       " hunt_name:%s]\n",
		       hunt[i].hunt_signal.signo,
		       hunt[i].hunt_signal.size,
		       hunt[i].owner,
		       hunt[i].hunt_name);
	}
}

static void print_tmo(int count, struct linx_info_tmo *tmo)
{
	int i;

	printf("\tTimeouts:\n");
	for (i = 0; i < count; i++) {
		printf("\t [tmoref:%#08x   %u signo:%-10d size:%6d]\n",
		       tmo[i].tmoref,
		       tmo[i].tmo,
		       tmo[i].tmo_signal.signo,
		       tmo[i].tmo_signal.size);
	}
}

static void print_filter(LINX_SIGSELECT *filter)
{
	int i;
	int count = abs(filter[0]) + 1;

	printf("\tFilters:\n");
	if (count == 1) {
		printf("\t recv[any]\n");
		return;
	}

	if ((int)(filter[0]) < 0)
		printf("\t recv but[ ");
	else
		printf("\t recv[ (%d) ", filter[0]);
	for (i = 1; i < count; i++) {
		printf("%x ", filter[i]);
	}
	printf("]\n");
}

static void print_signals(int count, struct linx_info_signal_2 *signals)
{
	int i;

	if (count > 0)
		printf("\tSignal queue:\n");
	else
		return;

	for (i = 0; i < count; i++) {
		printf("\t [position:%03d signo:%-10d "
		       "size:%-10d from:0x%08x oob:%s]\n",
		       i,
		       signals[i].signo,
		       signals[i].size,
		       signals[i].from,
		       signals[i].flags & MSG_OOB ? "yes" : "no");
	}
}

static void print_sock_stat(struct linx_info_stat *info_stat)
{
	printf("\tsent bytes:%llu recv bytes:%llu\n",
	       (long long unsigned int)info_stat->no_sent_bytes,
	       (long long unsigned int)info_stat->no_recv_bytes);

	printf("\tsent signals:%llu recv signals:%llu\n",
	       (long long unsigned int)info_stat->no_sent_signals,
	       (long long unsigned int)info_stat->no_recv_signals);

	printf("\tsent local signals:%llu recv local bytes:%llu\n",
	       (long long unsigned int)info_stat->no_sent_local_bytes,
	       (long long unsigned int)info_stat->no_recv_local_bytes);

	printf("\tsent local signals:%llu recv local signals:%llu\n",
	       (long long unsigned int)info_stat->no_sent_local_signals,
	       (long long unsigned int)info_stat->no_recv_local_signals);

	printf("\tsent remote bytes:%llu recv remote bytes:%llu\n",
	       (long long unsigned int)info_stat->no_sent_remote_bytes,
	       (long long unsigned int)info_stat->no_recv_remote_bytes);

	printf("\tsent remote signals:%llu recv remote signals:%llu\n",
	       (long long unsigned int)info_stat->no_sent_remote_signals,
	       (long long unsigned int)info_stat->no_recv_remote_signals);
	
	printf("\tqueued bytes:%llu queued signals:%llu\n",
	       (long long unsigned int)info_stat->no_queued_bytes,
	       (long long unsigned int)info_stat->no_queued_signals);
}

/* prints the date of one LINX process */
static void print_proc_info(struct linx_proc_info *proc_info,
			    struct linx_proc_details *pd)
{
	char *ptr;
	char tmp[ATTACH_STR_MAX + 1];
	int i, other_width, name_width, terminal_width, contents_width;

	printf("%6d ", proc_info->pid);
	printf("0x%-8x ", proc_info->spid);

	printf("%-7s ", type_to_str(proc_info->type));

	printf("%-6s ", state_to_str(proc_info->state));

	printf("%-5d ", proc_info->msg_queued);

	if (proc_info->attach_from > 99999)
		i = sprintf(tmp, "XXXXX/");
	else
		i = sprintf(tmp, "%d/", proc_info->attach_from);
	ptr = tmp + i;

	if (proc_info->attach_to > 99999)
		sprintf(ptr, "XXXXX");
	else
		sprintf(ptr, "%d", proc_info->attach_to);
	printf("%-11s ", tmp);

	printf("%-5d ", proc_info->hunts);

	printf("%-5d ", proc_info->timeouts);

	other_width = 60; /* Other fields take 60 characters */
	/* TOC takes by default 78 characters */
	contents_width = other_width + maxNameLength;
	terminal_width = get_terminal_width();
	if (terminal_width < contents_width) /* TOC wraps anyway, let it wrap*/
		name_width = proc_info->hunt_name_len;
	else
		name_width = terminal_width - other_width;
	if (name_width > maxNameLength)
		name_width = maxNameLength;
	ptr = malloc(name_width + 1);
	strncpy(ptr, proc_info->hunt_name, name_width + 1);
	ptr[name_width] = '\0';
	printf("%-*s", name_width, ptr);

	free(ptr);

	printf("\n");

	if (pd == NULL)
		return;

	if (pd->attach_from != NULL) {
		print_attach(pd->no_attach_from, pd->attach_from,
			     LINX_ATTACH_FROM);
	}
	if (pd->attach_to != NULL) {
		print_attach(pd->no_attach_to, pd->attach_to,
			     LINX_ATTACH_TO);
	}
	if (pd->hunt != NULL) {
		print_hunt(pd->no_hunt, pd->hunt);
	}
	if (pd->tmo != NULL) {
		print_tmo(pd->no_tmo, pd->tmo);
	}
	if (pd->filter != NULL) {
		print_filter(pd->filter);
	}
	if (pd->signals != NULL) {
		print_signals(pd->no_of_signals, pd->signals);
	}
	if(pd->socket_stats != NULL) {
		print_sock_stat(pd->socket_stats);
	}
}

int main(int argc, char *argv[])
{
	LINX *linx;
	struct linx_pid_list *pid_list = NULL;
	struct linx_proc_info *info;
	unsigned int i;
	int atts, hunts, tmos, sigs, local, remote, links, opt;
	int proc_type = 0, do_proc_type = 0;
	int batch = 0, simple = 0;
	int flags = 0;

	/* Check arguments. */
	while ((opt = getopt(argc, argv, "abfHhl:qsSt:Tn:") ) != -1)
	{
		switch (opt) {
		case 'a':
			flags |= LINX_DETAILS_ATTACH_ALL;
			break;
		case 'b':
			batch = 1;
			break;
		case 'f':
			flags |= LINX_DETAILS_FILTER;
			break;
		case 'l':
			maxNameLength = atol(optarg);
			if (maxNameLength < MINIMUM_NAME_LENGTH)
			    maxNameLength = MINIMUM_NAME_LENGTH;
			break;
		case 'h':
			usage();
			break;
		case 'H':
			flags |= LINX_DETAILS_HUNT;
			break;
		case 'q':
			flags |= LINX_DETAILS_QUEUE;
			break;
		case 't':
			proc_type = get_proc_type(optarg);
			do_proc_type = 1;
			break;
		case 's':
			simple = 1;
			break;
		case 'S':
			flags |= LINX_DETAILS_SOCKET_STATS;
			break;
		case 'T':
			flags |= LINX_DETAILS_TMO;
			break;
		case 'n':
			hunt_name = optarg;
			break;
		default:
			usage();
			exit(EXIT_FAILURE);
			break;
		}
	}

	linx = linx_open("linxstat", 0, NULL);
	if (linx == NULL) {
		printf("Could not open linx: %d\n", errno);
		exit(EXIT_FAILURE);
	}

	if (linx_get_pid_list(linx, &pid_list) == -1) {
		fprintf(stderr, "Could not get process list\n");
		linx_close(linx);
		exit(EXIT_FAILURE);
	}
	if (batch) {
		printf("%lu\n", pid_list->count);
		goto do_exit;
	}
	if ((!simple) && (hunt_name == NULL))
        {
		atts = hunts = tmos = sigs = local = remote = links = 0;
		for (i = 0; i < pid_list->count; i++) {
			info = &(pid_list->proc[i]);
			if (info->spid == LINX_ILLEGAL_SPID)
				continue;
			
			/* statistics */
			atts  += info->attach_from;
			hunts += info->hunts;
			tmos  += info->timeouts;
			sigs  += info->msg_queued;
			switch (info->type) {
			case LINX_TYPE_LOCAL:
				local++;
				break;
			case LINX_TYPE_LINK:
				links++;
				break;
			case LINX_TYPE_REMOTE:
				remote++;
				break;
			default:
				break;
			}
		}
		printf(" local:%d remote:%d, link:%d\n", local, remote, links);
		printf(" pending attach:%d\n", atts);
		printf(" pending hunt:%d\n", hunts);
		printf(" pending timeout:%d\n", tmos);
		printf(" queued signals:%d\n", sigs);

	/* new code that allows user to use -l maxNameLength to alter the
	// length of name field to be printed
	// use the * in the printf format to get the width of field size from
	// parameter list
	*/
                printf("\n   PID spid\t  type"
                       "    state  queue attach(f/t) hunt  tmo   %-*s\n", maxNameLength, "name");
	}

	for (i = 0; i < pid_list->count; i++) {
		struct linx_proc_details *pd = NULL;
		info = &(pid_list->proc[i]);
		if (info->spid == LINX_ILLEGAL_SPID)
			continue;
		if (do_proc_type && (info->type != proc_type))
			continue;

		if (flags) {
			linx_get_proc_details(linx, info->spid, flags, &pd);
		}
		if ((hunt_name != NULL) && (strcmp(hunt_name, info->hunt_name) != 0))
		{
			/* not the one we are looking for */
			continue;
		}
		print_proc_info(info, pd);

		if (pd != NULL)
			linx_free_proc_details(linx, &pd);
	}

do_exit:
	linx_free_pid_list(linx, &pid_list);
	linx_close(linx);

	exit(EXIT_SUCCESS);
}
